src/entities/trades/sushiswap/Sushiswap.ts

```diff
@@ -0,0 +1,291 @@
+import { BaseProvider } from '@ethersproject/providers'
+import { UnsignedTransaction } from '@ethersproject/transactions'
+import invariant from 'tiny-invariant'
+import fetch from 'node-fetch'
+
+import { ChainId, ONE, TradeType, _100 } from '../../../constants'
+import { Currency } from '../../currency'
+import { CurrencyAmount, Fraction, Percent, Price, TokenAmount } from '../../fractions'
+import { maximumSlippage as defaultMaximumSlippage } from '../constants'
+import { Trade } from '../interfaces/trade'
+import { TradeOptions } from '../interfaces/trade-options'
+import { RoutablePlatform } from '../routable-platform'
+import { getProvider, tryGetChainId, wrappedCurrency } from '../utils'
+import { SWAP_BASE_URL, getApiVersion } from './api'
+import { Contract } from 'ethers'
+import { SUSHISWAP_ROUTER_PROCESSOR_3_ABI } from './abi'
+import { parseUnits } from '@ethersproject/units'
+// import { approveAddressUrl, generateApiRequestUrl, RequestType } from './api'
+
+export interface SushiswapQuoteTypes {
+  amount: CurrencyAmount
+  quoteCurrency: Currency
+  tradeType: TradeType
+  maximumSlippage?: Percent
+  recipient?: string
+}
+
+interface SushiswapV3ConstructorParams {
+  maximumSlippage: Percent
+  inputAmount: CurrencyAmount
+  outputAmount: CurrencyAmount
+  tradeType: TradeType
+  chainId: ChainId
+  approveAddress: string
+  priceImpact: Percent
+  routeCode: string
+}
+
+export class SushiswapTrade extends Trade {
+  public readonly routeCode: string
+
+  public constructor({
+    maximumSlippage,
+    inputAmount,
+    outputAmount,
+    tradeType,
+    chainId,
+    approveAddress,
+    routeCode,
+    priceImpact,
+  }: SushiswapV3ConstructorParams) {
+    super({
+      details: undefined,
+      type: tradeType,
+      inputAmount,
+      outputAmount,
+      maximumSlippage,
+      platform: RoutablePlatform.SUSHISWAP,
+      chainId,
+      executionPrice: new Price({
+        baseCurrency: inputAmount.currency,
+        quoteCurrency: outputAmount.currency,
+        denominator: inputAmount.raw,
+        numerator: outputAmount.raw,
+      }),
+      priceImpact,
+      fee: new Percent('0', '1'),
+      approveAddress,
+    })
+    this.routeCode = routeCode
+  }
+
+  static async getQuote(
+    { amount, quoteCurrency, tradeType, maximumSlippage = defaultMaximumSlippage, recipient }: SushiswapQuoteTypes,
+    provider?: BaseProvider
+  ): Promise<SushiswapTrade | null> {
+    const chainId = tryGetChainId(amount, quoteCurrency)
+
+    if (!chainId) {
+      throw new Error('getQuote: chainId is required')
+    }
+
+    provider = provider || getProvider(chainId)
+
+    // Ensure the provider's chainId matches the provided currencies
+    invariant(
+      (await provider.getNetwork()).chainId == chainId,
+      `OneInch.getQuote: currencies chainId does not match provider's chainId`
+    )
+
+    const currencyIn = amount.currency
+    const currencyOut = quoteCurrency

```

Comment:
This is not true, the correct terms are:

`setToken` (the token that is set on the swapbox can be the in our out)
`quoteToken`(the token you want the quote)

Because the quote can be the currencyIn.

Reply:
In this case, Sushiswap API only accepts exact input, which means amount is always the currency amount in and quote is always out. In this context it suits well, but I agree if we have both cases, we should have a different name.